src/hugr/serialize.rs

```diff
@@ -89,120 +95,99 @@ impl<'de> Deserialize<'de> for Hugr {
 
-        let find_offset = |node: NodeIndex, offset: usize, dir: Direction| {
-            let sig = &op_types[node].signature();
+        let find_offset = |node: Node, offset: usize, dir: Direction, hugr: &Hugr| {
+            let sig = hugr.get_optype(node).signature();
             let offset = match offset < sig.port_count(dir) {
                 true => Some(offset as u16),
                 false => None,
             };
             (node, offset)
         };
 
-        let edges: Vec<_> = graph
-            .nodes_iter()
+        let edges: Vec<_> = hugr
+            .nodes()
             .flat_map(|node| {
-                graph
-                    .outputs(node)
+                hugr.node_ports(node, Direction::Outgoing)
                     .enumerate()
                     .flat_map(move |(src_offset, port)| {
-                        let src = find_offset(node, src_offset, Direction::Outgoing);
-                        graph.port_links(port).map(move |(_, tgt)| {
-                            let tgt_node = graph.port_node(tgt).unwrap();
-                            let tgt_offset = graph.port_offset(tgt).unwrap().index();
-                            let tgt = find_offset(tgt_node, tgt_offset, Direction::Incoming);
+                        let src = find_offset(node, src_offset, Direction::Outgoing, hugr);
+                        hugr.linked_ports(node, port).map(move |(tgt_node, tgt)| {
+                            let tgt = find_offset(
+                                tgt_node,
+                                tgt.offset.index(),
+                                Direction::Incoming,
+                                hugr,
+                            );
                             [src, tgt]
                         })
                     })
             })
             .collect();
 
-        Ok(Self {
-            nodes,
-            edges,
-            root: *root,
-        })
+        Ok(Self { nodes, edges })
     }
 }
 
 impl TryFrom<SerHugrV0> for Hugr {
     type Error = HUGRSerializationError;
-    fn try_from(SerHugrV0 { nodes, edges, root }: SerHugrV0) -> Result<Self, Self::Error> {
-        let mut hierarchy = Hierarchy::new();
-
+    fn try_from(SerHugrV0 { nodes, edges }: SerHugrV0) -> Result<Self, Self::Error> {
+        // Root must be first node
+        let mut nodes = nodes.into_iter();
         // if there are any unconnected ports or copy nodes the capacity will be
         // an underestimate
-        let mut graph = MultiPortGraph::with_capacity(nodes.len(), edges.len() * 2);
-        let mut op_types_sec = UnmanagedDenseMap::with_capacity(nodes.len());
-        for (parent, incoming, outgoing, typ) in nodes {
-            let ni = graph.add_node(incoming, outgoing);
-            if parent != ni {
-                hierarchy.push_child(ni, parent)?;
-            }
-            op_types_sec[ni] = typ;
+        let mut hugr = Hugr::with_capacity(nodes.next().unwrap().1, nodes.len(), edges.len() * 2);
```

Comment:
There could be a check for
```suggestion
        let root = nodes.next().unwrap();
        assert_eq!(root.0, NodeIndex::new(0).into());
        let mut hugr = Hugr::with_capacity(root.1, nodes.len(), edges.len() * 2);
```
(Or a new `HUGRSerializationError`)